staging/vhost-device-gpu/src/virtio_gpu.rs

```diff
@@ -0,0 +1,1011 @@
+    ///    2) To the host resource's attached iovecs
+    ///
+    /// Can also be used to invalidate caches.
+    fn transfer_read<'a>(
+        &mut self,
+        ctx_id: u32,
+        resource_id: u32,
+        transfer: Transfer3D,
+        buf: Option<VolatileSlice<'a>>,
+    ) -> VirtioGpuResult;
+
+    /// Attaches backing memory to the given resource, represented by a `Vec` of
+    /// `(address, size)` tuples in the guest's physical address space.
+    /// Converts to `RutabagaIovec` from the memory mapping.
+    fn attach_backing(
+        &mut self,
+        resource_id: u32,
+        mem: &GuestMemoryMmap,
+        vecs: Vec<(GuestAddress, usize)>,
+    ) -> VirtioGpuResult;
+
+    /// Detaches any previously attached iovecs from the resource.
+    fn detach_backing(&mut self, resource_id: u32) -> VirtioGpuResult;
+
+    /// Updates the cursor's memory to the given `resource_id`, and sets its
+    /// position to the given coordinates.
+    fn update_cursor(
+        &mut self,
+        resource_id: u32,
+        cursor_pos: VhostUserGpuCursorPos,
+        hot_x: u32,
+        hot_y: u32,
+    ) -> VirtioGpuResult;
+
+    /// Moves the cursor's position to the given coordinates.
+    fn move_cursor(&mut self, resource_id: u32, cursor: VhostUserGpuCursorPos) -> VirtioGpuResult;
+
+    /// Returns a uuid for the resource.
+    fn resource_assign_uuid(&self, resource_id: u32) -> VirtioGpuResult;
+
+    /// Gets rutabaga's capset information associated with `index`.
+    fn get_capset_info(&self, index: u32) -> VirtioGpuResult;
+
+    /// Gets a capset from rutabaga.
+    fn get_capset(&self, capset_id: u32, version: u32) -> VirtioGpuResult;
+
+    /// Creates a rutabaga context.
+    fn create_context<'a>(
+        &mut self,
+        ctx_id: u32,
+        context_init: u32,
+        context_name: Option<&'a str>,
+    ) -> VirtioGpuResult;
+
+    /// Get an `EventFd` descriptor, that signals when to call `event_poll`.
+    fn get_event_poll_fd(&self) -> Option<EventFd>;
+
+    /// Polls the Rutabaga backend.
+    fn event_poll(&self);
+}
+
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord)]
+pub enum VirtioGpuRing {
+    Global,
+    ContextSpecific { ctx_id: u32, ring_idx: u8 },
+}
+
+struct FenceDescriptor {
+    ring: VirtioGpuRing,
+    fence_id: u64,
+    desc_index: u16,
+    len: u32,
+}
+
+#[derive(Default)]
+pub struct FenceState {
+    descs: Vec<FenceDescriptor>,
+    completed_fences: BTreeMap<VirtioGpuRing, u64>,
+}
+
+#[derive(Copy, Clone, Debug, Default)]
+struct AssociatedScanouts(u32);
+
+impl AssociatedScanouts {
+    fn enable(&mut self, scanout_id: u32) {
+        self.0 |= 1 << scanout_id;
+    }
+
+    fn disable(&mut self, scanout_id: u32) {
+        self.0 ^= 1 << scanout_id;
+    }
+
+    const fn has_any_enabled(self) -> bool {
+        self.0 != 0
+    }
+
+    fn iter_enabled(self) -> impl Iterator<Item = u32> {
+        (0..VIRTIO_GPU_MAX_SCANOUTS).filter(move |i| ((self.0 >> i) & 1) == 1)
+    }
+}
```

Comment:
Could we add some unit tests on these? They are a bit black magic because they use bitwise ops on an integer, so there's no checks that they produce the expected results.

Reply:
Makes sense, I'll add some test.
If you want to get rid of the bit operations, we could use the bit-set crate, (no such thing in the standard library :roll_eyes:) , so another option could be to add the bit-set crate dependency, or just using `HashSet<u32>` from standard library which would work, but it seemed so wasteful for such a simple thing.

Reply:
I don't mind the operations! it's just that from one perspective it's not easy to follow if you haven't written the code. But it's not bad here, it's a small section like you said.

If you want to pull a dependency bitflags might be better, and we use it in other crates too. It has an iterator: <https://docs.rs/bitflags/latest/bitflags/iter/struct.Iter.html>

But I digress, I think unit testing (perhaps in inlined rustdocs) should be enough! No need to make it more complicated.

Reply:
Ok I wrote an excessive unit test to help you not having to write something yourself, feel free to remove/add stuff from it if you choose to use it:

```rust
diff --git a/staging/vhost-device-gpu/src/virtio_gpu.rs b/staging/vhost-device-gpu/src/virtio_gpu.rs
index 91c3b28..749eb58 100644
--- a/staging/vhost-device-gpu/src/virtio_gpu.rs
+++ b/staging/vhost-device-gpu/src/virtio_gpu.rs
@@ -1008,4 +1008,40 @@ mod tests {
             .unwrap_err();
         }
     }
+
+    #[test]
+    fn test_virtio_gpu_associated_scanouts() {
+        use super::AssociatedScanouts;
+
+        let mut scanouts = AssociatedScanouts::default();
+        assert!(!scanouts.has_any_enabled());
+        assert_eq!(scanouts.iter_enabled().next(), None);
+        scanouts.enable(1);
+        assert!(scanouts.has_any_enabled());
+        scanouts.disable(1);
+        assert!(!scanouts.has_any_enabled());
+        for scanout in 0..VIRTIO_GPU_MAX_SCANOUTS {
+            scanouts.enable(scanout);
+        }
+        assert!(scanouts.has_any_enabled());
+        assert_eq!(
+            scanouts.iter_enabled().collect::<Vec<u32>>(),
+            (0..VIRTIO_GPU_MAX_SCANOUTS)
+                .into_iter()
+                .collect::<Vec<u32>>()
+        );
+        for s in (0..VIRTIO_GPU_MAX_SCANOUTS).filter(|i| i % 2 == 0) {
+            scanouts.disable(s);
+        }
+        assert_eq!(
+            scanouts.iter_enabled().collect::<Vec<u32>>(),
+            (0..VIRTIO_GPU_MAX_SCANOUTS)
+                .filter(|i| i % 2 != 0)
+                .collect::<Vec<u32>>()
+        );
+        for s in (0..VIRTIO_GPU_MAX_SCANOUTS).filter(|i| i % 2 != 0) {
+            scanouts.disable(s);
+        }
+        assert!(!scanouts.has_any_enabled());
+    }
 }
```

Reply:
Btw,

> I think unit testing (perhaps in inlined rustdocs) 

Isn't possible because it's a private type.

Reply:
Thanks @epilys for the tests, I’ll review and add it. but I might tweak the naming or refactor some parts for clarity while I’m at it.

Reply:
@dorindabassey no problem, change whatever you want, this was just a rough sketch.